File: packages/adapters/src/ai.ts
/**
 * AI contract — conversational and agentic language-model calls.
 *
 * Mirrors `fiducial_adapters::ai` in Rust, which carries the full argument for
 * why this contract targets an **AI gateway** rather than one adapter per
 * model vendor. The short version: no honest intersection of the Anthropic and
 * OpenAI shapes exists, so the contract targets one shape — the gateway's —
 * and model choice becomes a string in a declaration.
 *
 * `openrouter` is real only here, the same boundary as `Turnstile` and
 * `SupabaseAuth`: a plain HTTPS API Rust could reach, with no Rust consumer
 * that would.
 */

/** Who authored a message in the conversation. */
export type Role = "system" | "user" | "assistant" | "tool";

/** A tool invocation the model asked for. */
export interface ToolCall {
  /** Gateway-assigned id. A `tool` message answers it by id, not by name — a
   * model may call the same tool twice in one turn. */
  id: string;
  name: string;
  /**
   * Arguments as a JSON **string**, not a parsed value.
   *
   * The caller has to validate these against the schema it declared either
   * way: a model can emit arguments that are valid JSON and still wrong. An
   * adapter that parsed them would be doing half of a job the caller cannot
   * skip, and would turn a model mistake into a thrown adapter error.
   */
  arguments: string;
}

/** One turn in the conversation. */
export interface Message {
  role: Role;
  /** Text content. Empty for an assistant turn that was only tool calls. */
  content: string;
  /** Set on a `tool` message: which call this answers. */
  toolCallId?: string;
  /** Set on an `assistant` turn replayed into a follow-up request. */
  toolCalls?: ToolCall[];
}

/** A tool the model may call. */
export interface ToolDefinition {
  name: string;
  /** What the tool does. The model reads this to decide whether to call it,
   * so it is part of the prompt, not documentation. */
  description: string;
  /** JSON Schema for the arguments. */
  parameters: Record<string, unknown>;
}

/** Whether the model must call a tool. Defaults to `"auto"` when tools are present. */
export type ToolChoice = "auto" | "none" | "required";

/** A request for a completion. */
export interface ChatRequest {
  messages: Message[];
  /**
   * System prompt, top-level.
   *
   * Top-level rather than a first `system` message because the two directions
   * are not equally easy: this lowers to a system message in one line, and
   * recovering "which of these was the system prompt" from a list does not
   * always work.
   */
  system?: string;
  tools?: ToolDefinition[];
  toolChoice?: ToolChoice;
  /**
   * Override the model for this one call.
   *
   * Normally omitted: the model is a declaration (`[ai] model` in
   * `fiducial.toml`) derived into the generated factory. This exists because
   * an agent that classifies with a small model and answers with a large one
   * is a normal agent — and making that product construct a second adapter
   * would turn the declaration into a lie rather than a default.
   */
  model?: string;
  maxTokens?: number;
  temperature?: number;
}

/** Why generation stopped. */
export type StopReason = "stop" | "length" | "tool_calls" | "other";

/** Tokens billed for one call. */
export interface Usage {
  inputTokens: number;
  outputTokens: number;
}

/** A completed response. */
export interface ChatResponse {
  text: string;
  /** Non-empty exactly when `stopReason` is `"tool_calls"`. */
  toolCalls: ToolCall[];
  stopReason: StopReason;
  usage: Usage;
  /**
   * The model that actually served the call.
   *
   * Reported rather than assumed: a gateway may route to a different model
   * than the one requested, and a product logging what it *asked for* is
   * logging a guess.
   */
  model: string;
}

/** One event in a streamed response. */
export type StreamEvent =
  /** A chunk of assistant text. Concatenated in order, these are `ChatResponse.text`. */
  | { type: "text"; text: string }
  /**
   * A complete tool call.
   *
   * Emitted whole, not as argument fragments: gateways stream tool arguments
   * as partial JSON, and partial JSON is not something a consumer can do
   * anything with except buffer it. The adapter buffers once so every
   * consumer does not.
   */
  | { type: "tool_call"; call: ToolCall }
  /** The final event. Always emitted on a successful stream. */
  | { type: "done"; stopReason: StopReason; usage: Usage };

/** What went wrong, in a form a caller can branch on. */
export type AiErrorKind =
  | "rejected"
  | "rate_limited"
  | "context_length_exceeded"
  | "model_unavailable"
  | "other";

/**
 * One error class with a `kind`, mirroring the Rust `AiError` enum.
 *
 * `context_length_exceeded` and `rate_limited` are distinguished from
 * `rejected` because their remedies are programmable — drop old turns and
 * retry; back off and retry — where the remedy for `rejected` is to read the
 * message.
 */
export class AiError extends Error {
  constructor(
    message: string,
    public readonly kind: AiErrorKind = "other",
    /** Set when `kind` is `"rate_limited"`, if the gateway said. */
    public readonly retryAfterSeconds?: number,
    /** Set when the error names a specific model. */
    public readonly model?: string,
  ) {
    super(message);
    this.name = "AiError";
  }
}

/** Conversational and agentic language-model calls, through a gateway. */
export interface Ai {
  /** Run a completion to the end and return it. */
  chat(request: ChatRequest): Promise<ChatResponse>;

  /**
   * Run a completion, yielding events as they arrive.
   *
   * A separate method rather than a `stream: boolean` on `ChatRequest`,
   * because a flag that changes what comes back is a return type in disguise.
   */
  stream(request: ChatRequest): AsyncIterable<StreamEvent>;
}

// ── None implementation ───────────────────────────────────────────────────────

const NONE_MESSAGE =
  'ai = "none": no AI vendor is selected. Set `[adapters] ai = "openrouter"` ' +
  "and `[ai] model` in fiducial.toml, then re-run `fid derive`.";

/**
 * No-op AI — every call throws `AiError` naming the fix.
 *
 * **The second `None*` that fails rather than succeeding**, after `NoneAuth`.
 * A no-op send or enqueue is indistinguishable from the real thing at the call
 * site: the caller wanted an effect elsewhere and does not read a result. A
 * completion *is* the result, so returning an empty one turns "no AI vendor is
 * selected" into a blank answer in the product's UI — which reads as a model
 * bug and gets debugged as one.
 *
 * It still costs nothing to wire in: constructing it is free and it sits in
 * `AdapterSet` from the first commit. Nothing fails until something actually
 * asks for a completion, at which point failing is the honest answer.
 */
export class NoneAi implements Ai {
  // Accepts and ignores `env` and the derived model so every vendor class in
  // this contract shares one constructor shape for the generated factory.
  constructor(_env?: unknown, _model?: string) {}

  async chat(_request: ChatRequest): Promise<ChatResponse> {
    throw new AiError(NONE_MESSAGE, "rejected");
  }

  async *stream(_request: ChatRequest): AsyncIterable<StreamEvent> {
    throw new AiError(NONE_MESSAGE, "rejected");
  }
}

// ── OpenRouter ────────────────────────────────────────────────────────────────

const OPENROUTER_URL = "https://openrouter.ai/api/v1/chat/completions";

/** The gateway's wire shape. Named here rather than imported — this package
 * depends on no vendor SDK, same as every other contract in it. */
interface WireToolCall {
  index?: number;
  id?: string;
  function?: { name?: string; arguments?: string };
}

interface WireChoice {
  message?: { content?: string | null; tool_calls?: WireToolCall[] };
  delta?: { content?: string | null; tool_calls?: WireToolCall[] };
  finish_reason?: string | null;
}

interface WireResponse {
  model?: string;
  choices?: WireChoice[];
  usage?: { prompt_tokens?: number; completion_tokens?: number };
  error?: { message?: string; code?: number | string };
}

function stopReasonFrom(finish: string | null | undefined): StopReason {
  switch (finish) {
    case "stop":
      return "stop";
    case "length":
      return "length";
    case "tool_calls":
    case "function_call":
      return "tool_calls";
    default:
      return "other";
  }
}

/**
 * Classify a failed response body into an `AiErrorKind`.
 *
 * Matched on the message text because the gateway multiplexes many upstream
 * providers and does not normalize these into distinct codes — a context
 * overflow arrives as a 400 whose message says so. Matching text is fragile,
 * so an unrecognized failure stays `"rejected"` (which every caller already
 * has to handle) rather than being guessed into a kind whose remedy would be
 * wrong.
 */
function classify(status: number, message: string): AiErrorKind {
  const m = message.toLowerCase();
  if (status === 429) return "rate_limited";
  if (m.includes("context length") || m.includes("maximum context")) {
    return "context_length_exceeded";
  }
  if (status === 404 || m.includes("no allowed providers") || m.includes("not a valid model")) {
    return "model_unavailable";
  }
  return "rejected";
}

/**
 * OpenRouter — a vendor-independent AI gateway speaking one normalized shape
 * across model providers.
 *
 * Secret-reached: `env.OPENROUTER_API_KEY`, set with `wrangler secret put`.
 * The constructor throws when the key is absent, with that command in the
 * message — a misconfigured product fails at construction rather than at the
 * first completion, which is the difference between a deploy-time error and a
 * user-facing one.
 *
 * `model` is the default the product declared in `[ai] model`; `fid derive`
 * passes it in. A per-call `request.model` wins over it.
 *
 * `fetchImpl` defaults to global `fetch` and exists so tests can pass a stub
 * rather than monkeypatching `globalThis.fetch`.
 */
export class OpenRouterAi implements Ai {
  private readonly apiKey: string;
  private readonly defaultModel: string;
  private readonly fetchImpl: typeof fetch;

  constructor(
    env: { OPENROUTER_API_KEY?: string },
    model?: string,
    fetchImpl: typeof fetch = fetch,
  ) {
    if (!env?.OPENROUTER_API_KEY) {
      throw new AiError(
        "OpenRouterAi: no `OPENROUTER_API_KEY` on env — set it with " +
          "`wrangler secret put OPENROUTER_API_KEY`",
        "rejected",
      );
    }
    if (!model) {
      throw new AiError(
        "OpenRouterAi: no model. Declare `[ai] model` in fiducial.toml and " +
          "re-run `fid derive`, or pass `model` on the request.",
        "model_unavailable",
      );
    }
    this.apiKey = env.OPENROUTER_API_KEY;
    this.defaultModel = model;
    this.fetchImpl = fetchImpl;
  }

  /** Translate the contract's request into the gateway's wire shape. */
  private body(request: ChatRequest, stream: boolean): Record<string, unknown> {
    const messages: Record<string, unknown>[] = [];
    if (request.system) {
      messages.push({ role: "system", content: request.system });
    }
    for (const m of request.messages) {
      const wire: Record<string, unknown> = { role: m.role, content: m.content };
      if (m.toolCallId) wire.tool_call_id = m.toolCallId;
      if (m.toolCalls?.length) {
        wire.tool_calls = m.toolCalls.map((c) => ({
          id: c.id,
          type: "function",
          function: { name: c.name, arguments: c.arguments },
        }));
      }
      messages.push(wire);
    }

    const body: Record<string, unknown> = {
      model: request.model ?? this.defaultModel,
      messages,
    };
    if (stream) body.stream = true;
    if (request.tools?.length) {
      body.tools = request.tools.map((t) => ({
        type: "function",
        function: {
          name: t.name,
          description: t.description,
          parameters: t.parameters,
        },
      }));
      body.tool_choice = request.toolChoice ?? "auto";
    }
    if (request.maxTokens !== undefined) body.max_tokens = request.maxTokens;
    if (request.temperature !== undefined) body.temperature = request.temperature;
    return body;
  }

  private async post(request: ChatRequest, stream: boolean): Promise<Response> {
    let res: Response;
    try {
      res = await this.fetchImpl(OPENROUTER_URL, {
        method: "POST",
        headers: {
          Authorization: `Bearer ${this.apiKey}`,
          "Content-Type": "application/json",
        },
        body: JSON.stringify(this.body(request, stream)),
      });
    } catch (err) {
      throw new AiError(`OpenRouterAi request failed: ${String(err)}`, "other");
    }

    if (!res.ok) {
      const text = await res.text().catch(() => "");
      let message = text;
      try {
        const parsed = JSON.parse(text) as WireResponse;
        if (parsed.error?.message) message = parsed.error.message;
      } catch {
        // Not JSON — the raw body is the best message available.
      }
      const kind = classify(res.status, message);
      const retryAfter = Number(res.headers?.get?.("retry-after") ?? "");
      throw new AiError(
        `OpenRouterAi returned HTTP ${res.status}: ${message || "(no body)"}`,
        kind,
        Number.isFinite(retryAfter) && retryAfter > 0 ? retryAfter : undefined,
        request.model ?? this.defaultModel,
      );
    }

    return res;
  }

  async chat(request: ChatRequest): Promise<ChatResponse> {
    const res = await this.post(request, false);
    const data = (await res.json()) as WireResponse;
    const choice = data.choices?.[0];

    // A 200 carrying an `error` is a real OpenRouter shape, not defensive
    // coding: an upstream provider failing mid-request is reported in the body
    // of an otherwise successful response.
    if (!choice) {
      throw new AiError(
        `OpenRouterAi: response had no choices${
          data.error?.message ? `: ${data.error.message}` : ""
        }`,
        classify(200, data.error?.message ?? ""),
      );
    }

    const toolCalls: ToolCall[] = (choice.message?.tool_calls ?? []).map((c) => ({
      id: c.id ?? "",
      name: c.function?.name ?? "",
      arguments: c.function?.arguments ?? "",
    }));

    return {
      text: choice.message?.content ?? "",
      toolCalls,
      stopReason: stopReasonFrom(choice.finish_reason),
      usage: {
        inputTokens: data.usage?.prompt_tokens ?? 0,
        outputTokens: data.usage?.completion_tokens ?? 0,
      },
      model: data.model ?? request.model ?? this.defaultModel,
    };
  }

  async *stream(request: ChatRequest): AsyncIterable<StreamEvent> {
    const res = await this.post(request, true);
    if (!res.body) {
      throw new AiError("OpenRouterAi: streamed response had no body", "other");
    }

    // Tool calls arrive as argument fragments keyed by `index`. Accumulate
    // them and emit each whole, once — see `StreamEvent`'s doc for why.
    const pending = new Map<number, { id: string; name: string; args: string }>();
    let stopReason: StopReason = "other";
    let usage: Usage = { inputTokens: 0, outputTokens: 0 };

    for await (const data of sseData(res.body)) {
      let parsed: WireResponse;
      try {
        parsed = JSON.parse(data) as WireResponse;
      } catch {
        // OpenRouter interleaves non-JSON keep-alive payloads; skipping one is
        // correct, and `[DONE]` never reaches here.
        continue;
      }

      if (parsed.error?.message) {
        throw new AiError(
          `OpenRouterAi stream error: ${parsed.error.message}`,
          classify(200, parsed.error.message),
        );
      }
      if (parsed.usage) {
        usage = {
          inputTokens: parsed.usage.prompt_tokens ?? 0,
          outputTokens: parsed.usage.completion_tokens ?? 0,
        };
      }

      const choice = parsed.choices?.[0];
      if (!choice) continue;

      const text = choice.delta?.content;
      if (text) yield { type: "text", text };

      for (const c of choice.delta?.tool_calls ?? []) {
        const index = c.index ?? 0;
        const slot = pending.get(index) ?? { id: "", name: "", args: "" };
        if (c.id) slot.id = c.id;
        if (c.function?.name) slot.name = c.function.name;
        if (c.function?.arguments) slot.args += c.function.arguments;
        pending.set(index, slot);
      }

      if (choice.finish_reason) stopReason = stopReasonFrom(choice.finish_reason);
    }

    for (const [, slot] of [...pending.entries()].sort((a, b) => a[0] - b[0])) {
      yield { type: "tool_call", call: { id: slot.id, name: slot.name, arguments: slot.args } };
    }

    yield { type: "done", stopReason, usage };
  }
}

/**
 * Yield the `data:` payloads of a server-sent-event stream, minus `[DONE]`.
 *
 * Written here rather than pulled in because it is twenty lines and the two
 * things that make SSE parsing wrong are both about buffering, not about
 * features: an event can straddle a chunk boundary, and a chunk can carry
 * several events. Splitting each chunk independently gets both wrong, and
 * gets them wrong only under load — which is when the bug is hardest to see.
 *
 * Comment lines (`:` — OpenRouter sends `: OPENROUTER PROCESSING` keep-alives)
 * are dropped, per the SSE spec.
 */
async function* sseData(body: ReadableStream<Uint8Array>): AsyncGenerator<string> {
  const decoder = new TextDecoder();
  const reader = body.getReader();
  let buffer = "";

  try {
    for (;;) {
      const { done, value } = await reader.read();
      if (done) break;
      buffer += decoder.decode(value, { stream: true });

      let newline: number;
      while ((newline = buffer.indexOf("\n")) !== -1) {
        const line = buffer.slice(0, newline).replace(/\r$/, "");
        buffer = buffer.slice(newline + 1);
        if (!line || line.startsWith(":")) continue;
        if (!line.startsWith("data:")) continue;
        const payload = line.slice("data:".length).trim();
        if (payload === "[DONE]") return;
        yield payload;
      }
    }
  } finally {
    reader.releaseLock();
  }
}

File: crates/fiducial-adapters/src/ai.rs
//! AI contract — conversational and agentic language-model calls.
//!
//! # Why this contract targets a gateway, not a model vendor
//!
//! The obvious shape for this contract was one adapter per model vendor —
//! `anthropic`, `openai`, `workers-ai` — behind their narrowest shared
//! interface. That shape does not work, and the reason is not a detail:
//!
//! - Anthropic puts `system` at the top level; OpenAI makes it a message role.
//! - Anthropic has content blocks; OpenAI has a string or a parts array.
//! - Streaming event shapes differ substantially between the two.
//! - The surface moves fast — thinking config, `tool_choice` and prefill all
//!   changed shape inside a year.
//!
//! Intersect those by hand and the contract is too thin to write an agent
//! against. Superset them and you have picked a vendor without admitting it,
//! which is `MISSION.md` principle 6 broken in the file that implements it.
//!
//! So the contract targets **one** shape — an AI gateway's — and *model
//! choice becomes a string in a declaration* rather than an adapter per
//! vendor. A gateway already does the cross-vendor normalization and, more
//! importantly, maintains it; vendor drift becomes the thing you are paying
//! the gateway for. `openrouter` is the first implementation.
//!
//! # Embeddings are deliberately absent
//!
//! Turning product data into vectors is a real second use, with its own
//! method shape, and it pairs with a vector store this platform does not
//! have. Adding `embed()` here before anything can store what it returns
//! would put a method on every implementation that no product can complete a
//! task with. It lands when it has a consumer.
//!
//! # Rust has the contract; `openrouter` is TypeScript-only
//!
//! Same boundary as `turnstile` and `supabase` (auth): a plain HTTPS API that
//! Rust could reach, with no Rust-side consumer that would use it. Every AI
//! call in the product shapes this platform scaffolds happens in a Worker or
//! a server route. The trait is defined here so the contract is one contract
//! in both languages — see `packages/adapters/src/ai.ts` for the half that
//! has a vendor behind it.

use thiserror::Error;

use crate::{BoxFuture, BoxStream};

/// Who authored a message in the conversation.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Role {
    /// Instructions for the model. Prefer [`ChatRequest::system`] — see its
    /// doc for why this variant still exists.
    System,
    User,
    Assistant,
    /// The result of a tool the assistant asked to call. Carries the
    /// `tool_call_id` it answers.
    Tool,
}

/// A tool invocation the model asked for.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ToolCall {
    /// Vendor-assigned id. A [`Role::Tool`] message answers it by id, not by
    /// name — a model may call the same tool twice in one turn.
    pub id: String,
    pub name: String,
    /// Arguments as a JSON **string**, not a parsed value.
    ///
    /// This crate is dependency-light on purpose and has no JSON type. More
    /// to the point, the caller has to validate these against the schema it
    /// declared regardless of who parses them: a model can emit arguments
    /// that are valid JSON and still wrong.
    pub arguments: String,
}

/// One turn in the conversation.
#[derive(Debug, Clone)]
pub struct Message {
    pub role: Role,
    /// Text content. Empty for an assistant turn that was only tool calls.
    pub content: String,
    /// Set on a [`Role::Tool`] message: which call this answers.
    pub tool_call_id: Option<String>,
    /// Set on a [`Role::Assistant`] message replayed back into a follow-up
    /// request: the calls that turn asked for.
    pub tool_calls: Vec<ToolCall>,
}

impl Message {
    /// A plain user turn — the common case, without five `None`s at the call
    /// site.
    pub fn user(content: impl Into<String>) -> Self {
        Self {
            role: Role::User,
            content: content.into(),
            tool_call_id: None,
            tool_calls: Vec::new(),
        }
    }

    /// A plain assistant turn.
    pub fn assistant(content: impl Into<String>) -> Self {
        Self {
            role: Role::Assistant,
            content: content.into(),
            tool_call_id: None,
            tool_calls: Vec::new(),
        }
    }

    /// The result of a tool call, answering it by id.
    pub fn tool_result(tool_call_id: impl Into<String>, content: impl Into<String>) -> Self {
        Self {
            role: Role::Tool,
            content: content.into(),
            tool_call_id: Some(tool_call_id.into()),
            tool_calls: Vec::new(),
        }
    }
}

/// A tool the model may call.
#[derive(Debug, Clone)]
pub struct ToolDefinition {
    pub name: String,
    /// What the tool does. The model reads this to decide whether to call it,
    /// so it is part of the prompt, not documentation.
    pub description: String,
    /// JSON Schema for the arguments, as a JSON string — same reason
    /// [`ToolCall::arguments`] is a string.
    pub parameters: String,
}

/// Whether the model must call a tool.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ToolChoice {
    /// The model decides. The default when tools are present.
    Auto,
    /// Tools are declared but must not be called this turn.
    None,
    /// The model must call one of the declared tools.
    Required,
}

/// A request for a completion.
#[derive(Debug, Clone, Default)]
pub struct ChatRequest {
    pub messages: Vec<Message>,
    /// System prompt, top-level.
    ///
    /// Carried top-level rather than as a first [`Role::System`] message
    /// because the two directions are not equally easy: a top-level system
    /// prompt lowers to a system message in one line, and recovering "which
    /// of these messages was the system prompt" from a list does not always
    /// work. Vendors that want a message role get one from the adapter.
    pub system: Option<String>,
    pub tools: Vec<ToolDefinition>,
    /// Defaults to [`ToolChoice::Auto`] when `tools` is non-empty.
    pub tool_choice: Option<ToolChoice>,
    /// Override the model for this one call.
    ///
    /// Normally `None`: the model is a declaration (`[ai] model` in
    /// `fiducial.toml`), derived into the generated factory and gated like
    /// every other derived fact. This exists because an agent that uses a
    /// small model for classification and a large one for the final answer is
    /// a normal agent, not an exotic one — and forcing that product to
    /// construct a second adapter would make the declaration a lie rather
    /// than a default.
    pub model: Option<String>,
    pub max_tokens: Option<u32>,
    pub temperature: Option<f32>,
}

/// Why generation stopped.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum StopReason {
    /// The model finished its turn.
    Stop,
    /// `max_tokens` was reached — the response is truncated.
    Length,
    /// The model is waiting on tool results.
    ToolCalls,
    /// Anything else the gateway reported, including content filtering.
    Other,
}

/// Tokens billed for one call.
#[derive(Debug, Clone, Copy, Default, PartialEq, Eq)]
pub struct Usage {
    pub input_tokens: u32,
    pub output_tokens: u32,
}

/// A completed response.
#[derive(Debug, Clone)]
pub struct ChatResponse {
    pub text: String,
    /// Non-empty exactly when `stop_reason` is [`StopReason::ToolCalls`].
    pub tool_calls: Vec<ToolCall>,
    pub stop_reason: StopReason,
    pub usage: Usage,
    /// The model that actually served the call.
    ///
    /// Reported rather than assumed: a gateway may route to a different model
    /// than the one requested (a fallback, a provider outage, an alias
    /// resolving to a dated snapshot), and a product logging what it *asked
    /// for* is logging a guess.
    pub model: String,
}

/// One event in a streamed response.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum StreamEvent {
    /// A chunk of assistant text. Concatenating every `Text` in order yields
    /// [`ChatResponse::text`].
    Text(String),
    /// A complete tool call.
    ///
    /// Emitted whole, not as argument fragments: gateways stream tool
    /// arguments as partial JSON, and a partial JSON string is not something
    /// a consumer can do anything with except buffer it. The adapter does
    /// that buffering once so every consumer does not.
    ToolCall(ToolCall),
    /// The final event. Always emitted on a successful stream.
    Done {
        stop_reason: StopReason,
        usage: Usage,
    },
}

/// Errors an [`Ai`] implementation may return.
#[derive(Debug, Error)]
pub enum AiError {
    /// The gateway refused the request — bad key, filtered content, malformed
    /// tool schema.
    #[error("request rejected by provider: {reason}")]
    Rejected { reason: String },
    #[error("rate limited — retry after {retry_after_secs}s")]
    RateLimited { retry_after_secs: u64 },
    /// The conversation is longer than the model's context window.
    ///
    /// Its own variant because the caller's remedy is specific and
    /// programmable — drop or summarize old turns and retry — where the
    /// remedy for [`AiError::Rejected`] is to read the message.
    #[error("context length exceeded for model `{model}`")]
    ContextLengthExceeded { model: String },
    /// The requested model does not exist, or the account cannot reach it.
    #[error("model `{model}` is not available: {reason}")]
    ModelUnavailable { model: String, reason: String },
    #[error("{0}")]
    Other(String),
}

/// Conversational and agentic language-model calls, through a gateway.
pub trait Ai: Send + Sync {
    /// Run a completion to the end and return it.
    fn chat<'a>(&'a self, request: &'a ChatRequest)
        -> BoxFuture<'a, Result<ChatResponse, AiError>>;

    /// Run a completion, yielding events as they arrive.
    ///
    /// A separate method rather than a flag on [`ChatRequest`] because the
    /// return types genuinely differ, and a `stream: bool` that changes what
    /// comes back is a return type in disguise.
    fn stream<'a>(
        &'a self,
        request: &'a ChatRequest,
    ) -> BoxStream<'a, Result<StreamEvent, AiError>>;
}

// ── None implementation ───────────────────────────────────────────────────────

/// No-op AI — every call fails with [`AiError::Rejected`] naming the fix.
///
/// **This is the second `None*` that fails rather than succeeding**, after
/// [`crate::NoneAuth`], and for a related reason. A no-op send ([`crate::NoneEmail`])
/// or a no-op enqueue is indistinguishable from the real thing at the call
/// site: the caller wanted an effect elsewhere and does not read a result. A
/// completion *is* the result. Returning an empty one turns "no AI vendor is
/// selected" into a blank answer in a product's UI, which reads as a model
/// bug and is debugged as one.
///
/// It still costs nothing to wire in: constructing it is free, it sits in
/// `AdapterSet` from the first commit, and nothing fails until something
/// actually asks for a completion — at which point failing is the honest
/// answer.
pub struct NoneAi;

const NONE_MESSAGE: &str = "ai = \"none\": no AI vendor is selected. Set \
                            `[adapters] ai = \"openrouter\"` and `[ai] model` \
                            in fiducial.toml, then re-run `fid derive`.";

impl Ai for NoneAi {
    fn chat<'a>(
        &'a self,
        _request: &'a ChatRequest,
    ) -> BoxFuture<'a, Result<ChatResponse, AiError>> {
        Box::pin(std::future::ready(Err(AiError::Rejected {
            reason: NONE_MESSAGE.to_string(),
        })))
    }

    fn stream<'a>(
        &'a self,
        _request: &'a ChatRequest,
    ) -> BoxStream<'a, Result<StreamEvent, AiError>> {
        Box::pin(Once(Some(Err(AiError::Rejected {
            reason: NONE_MESSAGE.to_string(),
        }))))
    }
}

/// A stream that yields one item and ends.
///
/// Hand-written rather than `futures_util::stream::once` so this crate depends
/// on `futures-core` — the `Stream` trait alone — and not on a combinator
/// library every consumer would then link. Six lines is cheaper than that.
struct Once<T>(Option<T>);

impl<T: Unpin> futures_core::Stream for Once<T> {
    type Item = T;

    fn poll_next(
        mut self: std::pin::Pin<&mut Self>,
        _cx: &mut std::task::Context<'_>,
    ) -> std::task::Poll<Option<T>> {
        std::task::Poll::Ready(self.0.take())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn ai() -> Box<dyn Ai> {
        Box::new(NoneAi)
    }

    fn request() -> ChatRequest {
        ChatRequest {
            messages: vec![Message::user("hello")],
            ..Default::default()
        }
    }

    #[tokio::test]
    async fn chat_fails_loudly_rather_than_returning_an_empty_completion() {
        let err = ai().chat(&request()).await.unwrap_err();
        let msg = err.to_string();
        assert!(msg.contains("no AI vendor is selected"), "{msg}");
        assert!(
            msg.contains("fid derive"),
            "the message names the fix: {msg}"
        );
    }

    /// The stream half fails the same way, and ends.
    ///
    /// A stream that errors and then hangs is worse than one that returns an
    /// error, so the termination is asserted, not just the error.
    #[tokio::test]
    async fn stream_fails_loudly_and_terminates() {
        use futures_util::StreamExt;

        let ai = ai();
        let req = request();
        let mut events = ai.stream(&req);
        let first = events.next().await.expect("one event");
        assert!(matches!(first, Err(AiError::Rejected { .. })));
        assert!(events.next().await.is_none(), "the stream must end");
    }

    #[test]
    fn none_ai_is_object_safe() {
        let _: Box<dyn Ai> = Box::new(NoneAi);
    }

    #[test]
    fn a_tool_result_answers_a_call_by_id_not_by_name() {
        let m = Message::tool_result("call_1", "{\"ok\":true}");
        assert_eq!(m.role, Role::Tool);
        assert_eq!(m.tool_call_id.as_deref(), Some("call_1"));
    }
}
